ext/http/00_serve.js

```diff
@@ -0,0 +1,462 @@
+import { InnerBody } from "ext:deno_fetch/22_body.js";
+import { Event } from "ext:deno_web/02_event.js";
+import { toInnerResponse } from "ext:deno_fetch/23_response.js";
+import { fromInnerRequest } from "ext:deno_fetch/23_request.js";
+import { AbortController } from "ext:deno_web/03_abort_signal.js";
+import {
+  _eventLoop,
+  _idleTimeoutDuration,
+  _idleTimeoutTimeout,
+  _protocol,
+  _readyState,
+  _rid,
+  _role,
+  _server,
+  _serverHandleIdleTimeout,
+  SERVER,
+  WebSocket,
+} from "ext:deno_websocket/01_websocket.js";
+import {
+  getReadableStreamResourceBacking,
+  readableStreamForRid,
+  ReadableStreamPrototype,
+  writableStreamForRid,
+} from "ext:deno_web/06_streams.js";
+const {
+  ObjectPrototypeIsPrototypeOf,
+  SafeSet,
+  SafeSetIterator,
+  SetPrototypeAdd,
+  SetPrototypeDelete,
+  Symbol,
+  TypeError,
+  Uint8ArrayPrototype,
+} = primordials;
+
+// Symbol we use to communicate to the rest of Deno that we handle upgrades ourselves
+const _wantsUpgrade = Symbol("_wantsUpgrade");
+const _upgraded = Symbol("_upgraded");
+
+class RequestHeaders {
+  #slabId;
+
+  get(header) {
+    const value = core.ops.op_get_request_header(this.#slabId, header);
+    return value;
+  }
+}
+
+class InnerRequest {
+  #slabId;
+  #headers;
+  #context;
+  #methodAndUri;
+  #streamRid;
+  #body;
+  #upgraded;
+
+  constructor(slabId, context) {
+    this.#slabId = slabId;
+    this.#context = context;
+    this.#upgraded = false;
+  }
+
+  close() {
+    if (this.#streamRid !== undefined) {
+      core.close(this.#streamRid);
+      this.#streamRid = undefined;
+    }
+    this.#slabId = undefined;
+  }
+
+  get [_upgraded]() {
+    return this.#upgraded;
+  }
+
+  [_wantsUpgrade](upgradeType, ...originalArgs) {
+    this.#upgraded = true;
+
+    // upgradeHttp is async
+    if (upgradeType == "upgradeHttp") {
+      core.ops.op_upgrade_http();
+    }
+
+    // upgradeHttpRaw is async
+    // TODO(mmastrac)
+    if (upgradeType == "upgradeHttpRaw") {
+      throw "";
+    }
+
+    // upgradeWebSocket is sync
+    if (upgradeType == "upgradeWebSocket") {
+      const response = originalArgs[0];
+      const ws = originalArgs[1];
+
+      this.url();
+      this.headerList;
+      this.close();
+
+      // Start the upgrade in the background.
+      (async () => {
+        //console.log(response.headerList);
+        const upgrade = await core.opAsync2(
+          "op_upgrade",
+          this.#slabId,
+          response.headerList,
+        );
+        const wsRid = core.ops.op_server_create(upgrade);
+        ws[_rid] = wsRid;
+        ws[_readyState] = WebSocket.OPEN;
+        ws[_role] = SERVER;
+        const event = new Event("open");
+        ws.dispatchEvent(event);
+
+        ws[_eventLoop]();
+        if (ws[_idleTimeoutDuration]) {
+          ws.addEventListener(
+            "close",
+            () => clearTimeout(ws[_idleTimeoutTimeout]),
+          );
+        }
+        ws[_serverHandleIdleTimeout]();
+      })();
```

Comment:
I believe there should be a `try/catch` here -- otherwise it will be an unhandled rejection if any of the APIs throws